Viewport hide objects

ericmehl opened the pull request:
This adds a new shortcut to the 3d viewer, Ctrl + H, to turn off visibility of the selected objects.

It also fixes a bug that left the enabled plug of some of the edit scope processors disconnected from their parent node. These should be connected to ensure disabled state is propagated.

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on python/GafferSceneUI/EditScopeUI.py:
```diff
@@ -101,3 +108,58 @@ def __pruningKeyPress( viewer, event ) :
+	if editScope is None or Gaffer.MetadataAlgo.readOnly( editScope ) :
+		return False
+
+	viewedNode = viewer.view()["in"].getInput().node()
+	if editScope != viewedNode and editScope not in Gaffer.NodeAlgo.upstreamNodes( viewedNode ) :
+		# Spare folks from hiding things in a downstream EditScope.
+		## \todo When we have a nice Viewer notification system we
+		# should emit a warning here.
+		return True
```
I think this is unnecessary - the inspector won't be editable if the EditScope isn't in the history.

ericmehl replied:
Right, inspectors will take care of that for us, fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5381/commits/ed1b6a582be422aa6c6a45587fda3bd794a284ba

johnhaddon commented on python/GafferSceneUI/EditScopeUI.py:
```diff
@@ -101,3 +108,58 @@ def __pruningKeyPress( viewer, event ) :
+
+	editScope = viewer.view().editScope()
+	if editScope is None or Gaffer.MetadataAlgo.readOnly( editScope ) :
+		return False
```
Should be `return True` I think. The `False` returns above are good, because they give other handlers a chance in the case where we have no intention of ever doing anything. But in this case, we'd like to do something but can't - I think it would be confusing to the user if we let other handlers do something different in this case.

ericmehl replied:
Got it, that makes sense. Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5381/commits/ed1b6a582be422aa6c6a45587fda3bd794a284ba

johnhaddon reviewed (changes requested):
Nice one - thanks Eric. No need to do the tweaks I've requested as fixup commits - it's all small enough that we might as well get it all squashed and ready to merge.
Cheers...
John

ericmehl commented:
Great! I fixed up the comments and squashed them down to https://github.com/GafferHQ/gaffer/pull/5381/commits/ed1b6a582be422aa6c6a45587fda3bd794a284ba. Should be good to go I think.

johnhaddon commented on Changes.md:
```diff
@@ -21,6 +21,7 @@ Improvements
 - OpenImageIOReader : The `availableFrames` plug no longer errors for a missing file sequence - instead it outputs an empty list.
 - PathListingWidget : Header data is now computed asynchronously, without locking the UI.
 - CollectScenes : Improved cancellation responsiveness for large lists of `rootNames`.
+- EditScopeUI : Added the ability to turn off visibility of selected objects in the viewer using the <kbd>Ctrl</kbd>+<kbd>H</kbd> shortcut.
```
This is in the section for the already-released 1.2.9.0, rather than the yet-to-be-released 1.2.x.x.

ericmehl replied:
Yep, little fella snuck in there without me noticing. Fixed in https://github.com/GafferHQ/gaffer/pull/5381/commits/833593820d6fc2cca4882c3e0d1784a640437c4d and https://github.com/GafferHQ/gaffer/pull/5381/commits/00188e8bb234c254769e30bc631c57a5781c8502

johnhaddon commented on python/GafferSceneUI/EditScopeUI.py:
```diff
@@ -101,3 +108,51 @@ def __pruningKeyPress( viewer, event ) :
+	with viewer.getContext() as context :
+		if (
+			not editScope["enabled"].getValue() or
+			( "AttributeEdits" in editScope and not editScope["AttributeEdits"]["enabled"].getValue() )
```
We shouldn't rely on the name of the processor here - there's nothing stopping a user modifying it. Instead we should use `EditScope.acquireProcessor( "AttributeEdits", createIfNecessary = False )`, which uses metadata instead.

johnhaddon reviewed (changes requested):
Sorry Eric, I just spotted one last thing that I should have seen the first time, and we also need to rejig Changes.md to account for the latest release. Details inline...

ericmehl commented:
A couple more tweaks taken care of and ready for a new look and hopefully merging!

johnhaddon commented:
Thanks!
